test/RelaySpec.hs

```diff
@@ -1,135 +1,740 @@
+          ownerSet (Root :/ bar1) [WireValue @Int32 1] $
+          define baz strDef $
+          postDefine baz pd $
+          trpdEmpty fooNs
+        verifyNoPdSub "owner" "sub" fooNs bazPdn
+        verifyNoTySub "owner" "sub" fooNs bazTn
+        verifyNoDataSub "owner" "sub" fooNs (Root :/ bar1)
+          [WireValue @Int32 42]
+
+    it "handles legitimate subscriptions" $ testRelay mempty $ do
+      -- Make an API with an owner:
+      sendFwd $ ClientConnect "Owner" "owner"
+      sendFwd $ ClientData "owner" $ Trpd $ simpleClaim foo
+      expect [emptyRootDig "owner", nsExists fooNs "owner"]
+
+      -- Connect a subscriber:
+      sendFwd $ ClientConnect "Subscriber" "sub"
+      expect [nsExists fooNs "sub"]
+
+      -- Subscribe to some data and check data and type subscriptions:
+      withSubscription fooNs root "sub" $ do
+        expect [ ServerData "sub" $ Frcud $ (frcudEmpty fooNs)
+          { frcudDefinitions = Map.singleton fooTn $ OpDefine intDef
+          , frcudTypeAssignments = Map.singleton Root (fooTn, Editable)
+          , frcudData = alSingleton Root $ ConstChange Nothing [WireValue @Int32 12]
+          }]
+        verifyDataSub "owner" "sub" fooNs Root [WireValue @Int32 13]
+        verifyTySub "owner" "sub" fooNs fooTn
+
+      -- After unsubscribing from the data, check that
+      -- a) The data subscription is terminated
+      verifyNoDataSub "owner" "sub" fooNs Root [WireValue @Int32 14]
+      -- b) The type subscription remains
+      verifyTySub "owner" "sub" fooNs fooTn
+
+      -- Ditch the residual type subscription
+      unsubscribe fooNs fooTn "sub"
+      verifyNoTySub "owner" "sub" fooNs fooTn
+
+      -- Test regulare sub/unsub cycle for Definitions
+      withSubscription fooNs fooTn "sub" $ do
+        expect [ ServerData "sub" $ Frcud $ (frcudEmpty fooNs)
+          { frcudDefinitions = Map.singleton fooTn $ OpDefine intDef }]
+        verifyTySub "owner" "sub" fooNs fooTn
+      verifyNoTySub "owner" "sub" fooNs fooTn
+
+      -- Test sub/unsub cycle for PostDefinitions too
+      let pd = postDef' "fooy" [(x, TtInt64 unbounded)]
+      sendFwd $ ClientData "owner" $ Trpd $ postDefine foo pd $ trpdEmpty fooNs
+      withSubscription fooNs fooPdn "sub" $ do
+        expect [ ServerData "sub" $ Frcud $ (frcudEmpty fooNs)
+         { frcudPostDefs = Map.singleton fooPdn $ OpDefine pd }]
+        verifyPdSub "owner" "sub" fooNs fooPdn
+      verifyNoPdSub "owner" "sub" fooNs fooPdn
+
+
+    it "should not send empty updates on empty subscriptions" $
+      testRelay mempty $ do
+        sendFwd $ ClientConnect "Subscriber" "sub"
+        expect [emptyRootDig "sub"]
+
+        sendFwd $ ClientData "sub" $ Trcsd $ mempty
+        nothingWaiting
+
+
+    it "is permissive/idempotent on client unsubscription" $ testRelay mempty $
+      -- The unsubscribed entity may have already gone away and the relay's
+      -- notifications crossed with the client's on the wire. The only case
```

Comment:
How does this combine with the comment above (about sending unsubs for things that are absent)?

Reply:
The above is questioning whether failed subscriptions and normally terminated subscriptions should look the same. This one is testing that unsubscription notifications only get sent once.